singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -741,6 +741,242 @@ class PTESolverRhoT : public mix_impl::PTESolverBase<EOSIndexer, RealIndexer> {
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::AssignIncrement;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::nmat;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::neq;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::ResidualNorm;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::vfrac_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::sie_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::eos;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rho;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::vfrac;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::sie;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::temp;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::press;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rho_total;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::uscale;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::utotal_scale;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::MatIndex;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::TryIdealPTE;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::jacobian;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::residual;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::dx;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::u;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::rhobar;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::Cache;
+  using mix_impl::PTESolverBase<EOSIndexer, RealIndexer>::Tnorm;
+
+ public:
+  // template the ctor to get type deduction/universal references prior to c++17
+  template <typename EOS_t, typename Real_t, typename Lambda_t>
+  PORTABLE_INLINE_FUNCTION
+  PTESolverPT(const int nmat, EOS_t &&eos, const Real vfrac_tot, const Real sie_tot,
+                Real_t &&rho, Real_t &&vfrac, Real_t &&sie, Real_t &&temp, Real_t &&press,
+                Lambda_t &&lambda, Real *scratch, const Real Tguess = 0.0)
+      : mix_impl::PTESolverBase<EOSIndexer, RealIndexer>(nmat, nmat + 1, eos, vfrac_tot,
+                                                         sie_tot, rho, vfrac, sie, temp,
+                                                         press, scratch, Tguess) {
+    vtemp = AssignIncrement(scratch, nmat);
+    // TODO(JCD): use whatever lambdas are passed in
+    /*for (int m = 0; m < nmat; m++) {
+      if (!variadic_utils::is_nullptr(lambda[m])) Cache[m] = lambda[m];
+    }*/
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real Init() {
+    InitBase();
+    Residual();
+    // calculate an initil equilibrium pressure
+    Real psum = 0.0;
+    Real vsum = 0.0;
+    for (int m = 0; m < nmat; ++m) {
+      psum += vfrac[m]*press[m];
+      vsum += vfrac[m];
+    }
+    // all normalized temps set to 1 so no averaging necessary here.
+    Tequil = temp[0];
+    Pequil = psum / vsum; 
```

Comment:
This machinery has always confused me. We pass in a `Tguess` that then gets saved as `Tnorm` (unless it isn't passed in) but it seems like we always use `temp[0]` as the initial starting point for the solver, yes?

We could take the same approach with `press[0]`, but your approach does provide a reasonable guess if the pressure is evolved by the host code in some way. We will just need to note that `press` needs to contain sane values of some sort.

While I'm at it, I'll create an issue to rename `Tguess` and note that `temp` requires at least _one_ sane value.

Reply:
See https://github.com/lanl/singularity-eos/issues/423

Reply:
Recall that `temp` is in arbitrary normalized temperature units and so it will always be multiplied by Tnorm.

Reply:
Don't we also normalize the pressures too?

Reply:
Yep, I think they are normalized by an internal energy.

Reply:
I'm fine with either duplicating the temperature approach (i.e. `Pequil = press[0]`) or staying with this. Whatever we do though, it needs to be noted in the documentation what values need to be sane.

Reply:
The pressures are initialized from a lookup so they are potentially different whereas there is 1 temperature guess calculated and it is folded into the temperature normalization and the temperature vector is initialized to 1 for every element.

Reply:
> The pressures are initialized from a lookup so they are potentially different

Can you point to this code?

Reply:
https://github.com/lanl/singularity-eos/blob/dholladay00/pt_solver/singularity-eos/closure/mixed_cell_models.hpp#L328

Reply:
Ah okay, I'm with you now. We should definitely document this.

To summarize my understanding:
- The initial temperature guess for the _PTE solver_ is `Tequil = temp[0]`
- The provided pressure can be nonsense since it is populated by `InitBase()`
- `InitBase()` calculates the initial pressure from `Tscale` and the input density, **_NOT_** `Tequil`
- You're volume fraction averaging those pressures to get `Pequil`, which is the initial PTE state